src/Datastore/Query/Query.php

```diff
@@ -0,0 +1,376 @@
+ *         ],
+ *         'filter' => [
+ *             'propertyFilter' => [
+ *                 'op' => 'EQUAL',
+ *                 'property' => [
+ *                     'name' => 'firstName'
+ *                 ],
+ *                 'value' => [
+ *                     'stringValue': 'Bob'
+ *                 ]
+ *             ]
+ *         ]
+ *     ]
+ * ]);
+ *
+ * $result = $datastore->runQuery($query);
+ * ```
+ *
+ * @see https://cloud.google.com/datastore/reference/rest/v1/projects/runQuery#query Query Object Reference
+ * @see https://cloud.google.com/datastore/docs/concepts/queries Datastore Queries
+ */
+class Query implements QueryInterface
+{
+    use DatastoreTrait;
+
+    private $allowedOperators = [
+        self::OP_LESS_THAN,
+        self::OP_LESS_THAN_OR_EQUAL,
+        self::OP_GREATER_THAN,
+        self::OP_GREATER_THAN_OR_EQUAL,
+        self::OP_EQUALS,
+        self::OP_HAS_ANCESTOR,
+    ];
+
+    private $allowedOrders = [
+        self::ORDER_ASCENDING,
+        self::ORDER_DESCENDING
+    ];
+
+    /**
+     * @var string
+     */
+    private $projectId;
+
+    /**
+     * @var array
+     */
+    private $options;
+
+    /**
+     * @param string $projectId The Google Cloud Platform Project ID
+     * @param array $options {
+     *     Configuration Options
+     *
+     *     @type array $query [Query](https://cloud.google.com/datastore/reference/rest/v1/projects/runQuery#query)
+     * }
+     */
+    public function __construct($projectId, array $options = [])
+    {
+        $this->projectId = $projectId;
+        $this->options = $options + [
+            'query' => [
+                'projection' => [],
+                'kind' => [],
+                'order' => [],
+                'distinctOn' => []
+            ]
+        ];
+    }
+
+    /**
+     * Indicate that this type does support automatic pagination.
+     *
+     * @access private
+     * @return bool
+     */
+    public function canPaginate()
+    {
+        return true;
+    }
+
+    /**
+     * Return a service-compliant array.
+     *
+     * This method is intended for use internally by the PHP client.
+     *
+     * @access private
+     * @return array
+     */
+    public function queryObject()
+    {
+        return array_filter($this->options['query']);
+    }
+
+    /**
+     * Return the query_type union field name.
+     *
+     * @return string
+     * @access private
+     */
+    public function queryKey()
+    {
+        return "query";
+    }
+
+    /**
+     * @access private
+     */
+    public function jsonSerialize()
+    {
+        return $this->queryObject();
+    }
+
+    /**
+     * Set the Query Projection.
+     *
+     * Accepts an array of properties. If set, only these properties will be
+     * returned.
+     *
+     * @param array|string $properties The property or properties to include in
+     *        the result.
+     * @return Query
+     */
+    public function projection($properties)
```

Comment:
The publicly accessible methods that will end up in our docs should all have examples